### PR TITLE
Added CIOU loss function

### PR DESCRIPTION
**Pull request by abhi-glitchhg**
#2980 

[Reference](https://github.com/facebookresearch/detectron2/blob/221448e4dcfbebf215b8d21ae7e4b1dfbf422d29/detectron2/layers/losses.py#L66) 

cc: @oke-aditya  @datumbox

### CODE REVIEW

**Comment by oke-aditya** [comment]
Hi. It's late in India rn. Will review it in daytime.

File: torchvision/ops/ciou_loss.py
Comment (datumbox) [inline torchvision/ops/ciou_loss.py:40]: We shouldn't put an assertion here to avoid slowdowns due to GPU<=>CPU syncs.

**Review by datumbox — COMMENTED** [review]
@abhi-glitchhg Thanks for the contribution!

I've added a few comments for your attention.

File: torchvision/ops/ciou_loss.py
Comment (oke-aditya) [inline torchvision/ops/ciou_loss.py:16]:
```suggestion
    Some description about CIoU
    This loss is symmetric, so the boxes1 and boxes2 arguments are interchangeable.
     Both sets of boxes are expected to be in ``(x1, y1, x2, y2)`` format with
    ``0 <= x1 < x2`` and ``0 <= y1 < y2``, and The two boxes should have the
    same dimensions.
```

**Review by oke-aditya — COMMENTED** [review]
Nice a few comments though

File: torchvision/ops/ciou_loss.py
Comment (datumbox) [reply torchvision/ops/ciou_loss.py:6]: Good spot. The name of the file and the method can't be the same. Else it will create ambiguities on the imports from `torchvision.ops`. Following the example of giou, the file could be called `ciou_loss.py` and the method `complete_box_iou_loss`; I'm not terribly happy with the latter if I'm honest. Thoughts?

File: torchvision/ops/ciou_loss.py
Comment (abhi-glitchhg) [reply torchvision/ops/ciou_loss.py:40]: Okay I will do that.

Comment (abhi-glitchhg) [reply torchvision/ops/ciou_loss.py:40]: A doubt - 

So the user has to manually check if there are no bad inputs to this function, right?  should we have some warning message in that scenario?

File: torchvision/ops/ciou_loss.py
Comment (abhi-glitchhg) [reply torchvision/ops/ciou_loss.py:33]: Thanks, Aditya for the review.  I love your feedback. :)
 I am not much familiar with jit logging. 
 
 Is there any material/documentation that I can go through to understand this? 
 Thanks.

File: torchvision/ops/ciou_loss.py
Comment (abhi-glitchhg) [reply torchvision/ops/ciou_loss.py:40]: Understood thanks!

File: torchvision/ops/ciou_loss.py
Comment (oke-aditya) [reply torchvision/ops/ciou_loss.py:33]: * PyTorch note: https://pytorch.org/docs/stable/notes/large_scale_deployments.html#api-usage-logging;
    * Logging policy: https://github.com/pytorch/vision/issues/5052;

also see https://github.com/pytorch/vision/blob/main/torchvision/utils.py#L525

File: torchvision/ops/ciou_loss.py
Comment (datumbox) [reply torchvision/ops/ciou_loss.py:40]: The problem is you can't have a warning unless you check and if you check you slowdown the execution significantly. The user needs to make sure that they don't provide bad inputs. If they want the warnings the recommendation is to wrap this method on their end and perform those checks.

You can see similar discussions at #5322, #5691 and #5684.

Comment (datumbox) [reply torchvision/ops/ciou_loss.py:40]: Damn, Github **really** needs to implement a feature to info you that others are writing comments at the same time. It happens all the time that we are posting concurrently the same things. lol :D

**Comment by yassineAlouini** [comment]
Working on dIoU, I have noticed that many things can be refactored between gIoU, cIoU, and dIoU.
I will push my contribution anyway but we can discuss what can be done next to improve things. :ok_hand:

**Comment by oke-aditya** [comment]
yeah :) There is a neat relation between these 3 and code can be re-used. :smile:

**Comment by yassineAlouini** [comment]
> yeah :) There is a neat relation between these 3 and code can be re-used. smile

Yes, I guess we can repeat some of the code for now since we have two branches at the same time + the existing gIoU. Next, we can start a new PR to refactor things a bit.

File: test/test_ops.py
Comment (oke-aditya) [inline test/test_ops.py:1455]: hmm, to be on safer side can you add `dtype=torch.float`. It might just take these as ints
also 
```suggestion
    box1 = ([-1., -1., 1., 1.],[-1., -1., 1., 1.],[-1., -1., 1., 1.],[-1., -1., 1., 1.])
```

File: torchvision/ops/ciou_loss.py
Comment (oke-aditya) [inline torchvision/ops/ciou_loss.py:9]: (but why duplicate?)

File: torchvision/ops/ciou_loss.py
Comment (oke-aditya) [inline torchvision/ops/ciou_loss.py:23]:
```suggestion
    """
    Need one line describing CIoU loss.
```

File: test/test_ops.py
Comment (abhi-glitchhg) [reply test/test_ops.py:1455]: oh yeah, will add tests with different datatypes.

**Comment by oke-aditya** [comment]
Please add docs in docs/ops.rst

File: test/test_ops.py
Comment (oke-aditya) [inline test/test_ops.py:82]: I think this is unrelated or due to formatter. But I think the original one is right.

File: test/test_ops.py
Comment (oke-aditya) [inline test/test_ops.py:118]: Same here

File: test/test_ops.py
Comment (oke-aditya) [inline test/test_ops.py:1593]:
```suggestion
            expected_output = torch.tensor(expected_output, device=device)
```

This will automatically take the dtype. You don't need to parameterize here.

File: test/test_ops.py
Comment (oke-aditya) [inline test/test_ops.py:1597]: 0.0 is treated with dtype as float and will fail when you pass boxes as torch.half. That's what CI us complaining.

File: torchvision/ops/ciou_loss.py
Comment (oke-aditya) [inline torchvision/ops/ciou_loss.py:84]:
```suggestion
    v = (4 / (torch.pi ** 2)) * torch.pow((torch.atan(w_gt / h_gt) - torch.atan(w_pred / h_pred)), 2)
```

Comment (oke-aditya) [reply torchvision/ops/ciou_loss.py:84]: We do have torch.pi. Very debated but finally included 

```
>>> torch.pi
3.141592653589793
```

the reason is general compatibility of array API (Broad topic, I have 0 knowledge) 
see
https://github.com/pytorch/pytorch/blob/main/torch/__init__.py#L587

File: torchvision/ops/ciou_loss.py
Comment (oke-aditya) [inline torchvision/ops/ciou_loss.py:1]: safely drop this import after you remove math.pi

```suggestion

```

File: test/test_ops.py
Comment (abhi-glitchhg) [reply test/test_ops.py:1597]: Ohh! 
Thanks for the help!

File: test/test_ops.py
Comment (abhi-glitchhg) [reply test/test_ops.py:82]: Tests were showing ufmt errors for `test/test_ops.py` file, so I used ufmt format for this file. 
Will revert these changes

Comment (oke-aditya) [reply test/test_ops.py:82]: I'm not sure, of formatting. Since, I use pre-commit and it automatically (auto magically?) corrects stuff and then safely commits. You might as well try it :)

**Comment by oke-aditya** [comment]
Once CI goes green (which probably should) please UNmark this as draft. @abhi-glitchhg

File: torchvision/ops/ciou_loss.py
Comment (abhi-glitchhg) [reply torchvision/ops/ciou_loss.py:1]: There exists functions which depend on `math` module. 

https://github.com/pytorch/vision/blob/467b841ac2650c962952bbee5a542edb835c98bc/test/test_ops.py#L277-L278

But i see no problem in changing these  from `math.floor()` to `torch.floor()`. 
@oke-aditya

Comment (oke-aditya) [reply torchvision/ops/ciou_loss.py:1]: I'm pointing to different file I guess. 
But good idea to replace there as well.

**Comment by datumbox** [comment]
@abhi-glitchhg @oke-aditya Thank you very much for your contributions and reviews so far.

Apologies for the delayed response, but I was OOO. I might be missing something but what is the reason we have duplicate implementations for cIoU in the loss and the box version? Is there a reason we don't share some of the code?

**Comment by yassineAlouini** [comment]
> @abhi-glitchhg @oke-aditya Thank you very much for your contributions and reviews so far.
> 
> Apologies for the delayed response, but I was OOO. I might be missing something but what is the reason we have duplicate implementations for cIoU in the loss and the box version? Is there a reason we don't share some of the code?

As far as my exploration goes, it seems it is the case for gIoU as well and I haven't found any reason why this should be the case but maybe I am missing something. I have also duplicated some code in dIoU. Perhaps it is best to have a PR to refactor some of the shared code between gIoU, cIoU, and dIoU?

**Comment by datumbox** [comment]
@yassineAlouini I think sending a separate PR to refactor gIoU makes a lot of sense. For new code (cIoU and dIoU) I would fix in the current open PRs.

**Comment by yassineAlouini** [comment]
@datumbox Noted, I will work on this (when I have allocated time) and update the code. :ok_hand:

File: torchvision/ops/boxes.py
Comment (datumbox) [inline torchvision/ops/boxes.py:339]: No need to upcast now. You've done it once above:
```suggestion
    whi = (rbi - lti).clamp(min=0)  # [N,M,2]
```

File: torchvision/ops/boxes.py
Comment (datumbox) [inline torchvision/ops/boxes.py:348]:
```suggestion
    centers_distance_squared = (x_p - x_g) ** 2 + (y_p - y_g) ** 2
```

**Review by datumbox — COMMENTED** [review]
@abhi-glitchhg Just a minor comment below. When you are ready for us to check the PR, remove the draft flag. :)

File: torchvision/ops/ciou_loss.py
Comment (datumbox) [inline torchvision/ops/ciou_loss.py:4]: Nit: This _upcast needs to move in `._utils`

**Review by datumbox — COMMENTED** [review]
@abhi-glitchhg  Thanks for the contribution. 

I'm looping @yassineAlouini who is looking at dIoU as many of the comments below are applicable.

I've verified the implementation both by reviewing and by comparing the two methods as follows:
```python
import torch

from torchvision.ops.boxes import complete_box_iou
from torchvision.ops.ciou_loss import complete_box_iou_loss

def random_box(canvas_size):
    x1y1 = torch.rand((1, 2)) * canvas_size
    wh = torch.rand((1, 2)) * canvas_size
    x2y2 = (x1y1 + wh).clamp(0, canvas_size)
    return torch.cat((x1y1, x2y2), axis=1)


canvas_size = 1000


for _ in range(10000):
    box1 = random_box(canvas_size)
    box2 = random_box(canvas_size)

    v1 = 1 - complete_box_iou(box1, box2)
    v2 = complete_box_iou_loss(box1, box2)

    torch.testing.assert_close(v1[0], v2, rtol=0, atol=1e-6)
```

This confirms that both implementations are equivalent. Though I think the PR is correct, I feel there is need for deep cleaning both for dIoU and cIoU (and gIoU). In particular we should consider refactoring the code across losses and boxes to share some of the internal implementations. Not everything is shareable and we should be mindful to maintain the speed but we shouldn't have duplicate implementations. Moreover, @pmeier suggested a few improvements on the test side at https://github.com/pytorch/vision/pull/5786#discussion_r857325564, so these improvements should also be completed.

Kindly let me know what direction you would like us to take now. We could bring both PRs on a similar state as this one and merge and then 1 of you can pick up the cleaning bit. Alternatively we can keep iterating on the current PRs until the nits are taken care of. It's up to you and I'm happy to take your preferred approach, I just want to make sure the implementations will be cleaned up prior the next release.

**Comment by abhi-glitchhg** [comment]
> Kindly let me know what direction you would like us to take now. We could bring both PRs on a similar state as this one and merge and then 1 of you can pick up the cleaning bit. Alternatively we can keep iterating on the current PRs until the nits are taken care of. It's up to you and I'm happy to take your preferred approach, I just want to make sure the implementations will be cleaned up prior the next release.

I would prefer the first approach(bring both PRs to similar state, merge, and then do refactor) but I'm also open to the second approach.

**Review by datumbox — APPROVED** [review]
@abhi-glitchhg Sounds good to me. Approving to unblock. We should do the clean ups once dIoU is merged.

@pmeier Any concerns/blocking changes on your side for tests?

**Review by pmeier — COMMENTED** [review]
Minor style nit, but otherwise testing looks good to me minus the things we discussed for the follow up PR.

**Comment by abhi-glitchhg** [comment]
Hey, @datumbox , what's the expected date for next release?

**Comment by pmeier** [comment]
Tentative date for branch cut, i.e. "feature freeze", is on 19th of May. After that it will be roughly one month until release, but no official date is set yet.

**Pull request merged**
